### PR TITLE
Fix exit_group test

### LINKED ISSUE #1937: [LTP] accept02 from LTP fails, but is not reported
## Description of the problem
`accept02` from LTP test suite fails, but `make regression` (hence CI) does not report it.

## Steps to reproduce
Run `./pal_loader ./accept02 ` in `LibOS/shim/test/ltp/install/testcases/bin` - it does not work:
`TBROK: setsockopt() failed unexpectedly: ENOPROTOOPT (92)`
Run `make regression` in `LibOS/shim/test/ltp/` - the above failure is not reported.

### PR DESCRIPTION
**Pull request by pwmarcz**
## Description of the changes 

~This fixes #1937, but more importantly, makes sure that any thread calling `exit()` will cause the correct exit code to be reported.~

The issue has been fixed by threads rework in #1949, this adds a test for `exit()` from a thread.

## How to test this PR? 

The `exit_group` test is a test for this scenario:

```sh
cd LibOS/shim/test/regression
make exit_group exit_group.manifest

# run the test:
../../../../Scripts/run-pytest -k test_401_exit_group -v

# trigger the binary manually:
./exit_group 3 100; echo $?
./pal_loader exit_group 3 100; echo $?
```

---
This change is [Reviewable](https://reviewable.io/reviews/oscarlab/graphene/1943)

### CODE REVIEW

**Review by yamahata — DISMISSED** [review]
> Reviewed 8 of 8 files at r1.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/1943)* status: all files reviewed, all discussions resolved, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: Intel)

**Review by boryspoplawski — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/1943)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @pwmarcz)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/1943#-MLJ85OjAjMbb4as_iF5:-MLJ85Ol5b9pZHCFgQ8s:b-sqlv36) (no related file):*
Today I'll submit a PR reworking threads (just waiting for #1938) which will get rid of these per thread exit codes and unfortunately remove all code introduced in this PR.
Now I know why I hit that LTP `accept02` issue :)

We could merge this as temporary fix, but I'll revert all this changes anyway. What you guys think? cc @dimakuv @mkow @yamahata 

---

**Review by pwmarcz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/1943)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @dimakuv, @mkow, and @yamahata)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/1943#-MLJ85OjAjMbb4as_iF5:-MLJAuaFCB2Wq2wXWDKB:b-yud1rg) (no related file):*
<details><summary><i>Previously, boryspoplawski (Borys Popławski) wrote…</i></summary><blockquote>

Today I'll submit a PR reworking threads (just waiting for #1938) which will get rid of these per thread exit codes and unfortunately remove all code introduced in this PR.
Now I know why I hit that LTP `accept02` issue :)

We could merge this as temporary fix, but I'll revert all this changes anyway. What you guys think? cc @dimakuv @mkow @yamahata 
</blockquote></details>

No problem, the change is somewhat hacky anyway, but you can still use the test :)

---

**Review by boryspoplawski — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/1943)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @dimakuv, @mkow, and @yamahata)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/1943#-MLJ85OjAjMbb4as_iF5:-MLJBSDo0ZoBLbhdoPnT:b-x6q5ls) (no related file):*
<details><summary><i>Previously, pwmarcz (Paweł Marczewski) wrote…</i></summary><blockquote>

No problem, the change is somewhat hacky anyway, but you can still use the test :)
</blockquote></details>

Yeah, we will definitely keep the reworked test, the question is when do we merge it. I expect threads rework merge to be lengthy process, as the PR will be quite big.

---

**Review by mkow — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/1943)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @dimakuv and @yamahata)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/1943#-MLJ85OjAjMbb4as_iF5:-MLJOw10BOniBsyFVX2T:bzhhbjz) (no related file):*
<details><summary><i>Previously, boryspoplawski (Borys Popławski) wrote…</i></summary><blockquote>

Yeah, we will definitely keep the reworked test, the question is when do we merge it. I expect threads rework merge to be lengthy process, as the PR will be quite big.
</blockquote></details>

@boryspoplawski Your call, I'm fine with both options. But if we decide to merge it, then I'd do it before submitting your changes, because they will conflict.

---

**Review by boryspoplawski — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/1943)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @dimakuv and @yamahata)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/1943#-MLJ85OjAjMbb4as_iF5:-MLJRlT380nbOcg_kXON:bes4o3a) (no related file):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

@boryspoplawski Your call, I'm fine with both options. But if we decide to merge it, then I'd do it before submitting your changes, because they will conflict.
</blockquote></details>

Given that it would be me, who needs to resolve the conflicts and that this is not a fatal issue, I would wait and merge just the test after threads PR :)

---

**Review by dimakuv — CHANGES_REQUESTED** [review]
> Reviewed 4 of 8 files at r1.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/1943)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @yamahata)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/1943#-MLJ85OjAjMbb4as_iF5:-MLM_tW07hZisx3kpbov:b6nhuzs) (no related file):*
<details><summary><i>Previously, boryspoplawski (Borys Popławski) wrote…</i></summary><blockquote>

Given that it would be me, who needs to resolve the conflicts and that this is not a fatal issue, I would wait and merge just the test after threads PR :)
</blockquote></details>

What about we drop the second commit from this PR (bar that skip of the `accept02` LTP test)? And mark it as blocked until we merge Borys's PR (`This PR must be rebased on top of ...`).

---

**Review by boryspoplawski — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/1943)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @yamahata)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/1943#-MLJ85OjAjMbb4as_iF5:-MLNipV-41LuHGphrq2g:b-yf4532) (no related file):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

What about we drop the second commit from this PR (bar that skip of the `accept02` LTP test)? And mark it as blocked until we merge Borys's PR (`This PR must be rebased on top of ...`).
</blockquote></details>

I guess we can live without this test (old version) in the meantime, so I'm ok with this

---

**Review by boryspoplawski — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/1943)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @yamahata)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/1943#-MLJ85OjAjMbb4as_iF5:-MMX53_g0z6HPIDBqmkz:bgyqlhi) (no related file):*
<details><summary><i>Previously, boryspoplawski (Borys Popławski) wrote…</i></summary><blockquote>

I guess we can live without this test (old version) in the meantime, so I'm ok with this
</blockquote></details>

Threads PR was merged, this can be updated (just the test will remain, I guess?)

---

**Review by pwmarcz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/1943)* status: 0 of 3 files reviewed, 1 unresolved discussion, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @dimakuv and @yamahata)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/1943#-MLJ85OjAjMbb4as_iF5:-MMZmcZCE9fR_FoJaLsb:b-hae5c8) (no related file):*
<details><summary><i>Previously, boryspoplawski (Borys Popławski) wrote…</i></summary><blockquote>

Threads PR was merged, this can be updated (just the test will remain, I guess?)
</blockquote></details>

Done. I rebased the code on newest master and left only the test.

---

**Review by dimakuv — CHANGES_REQUESTED** [review]
> Reviewed 3 of 8 files at r2.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/1943)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @yamahata)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/1943#-MLJ85OjAjMbb4as_iF5:-MMZxIJa6Nassumy3e_X:b4zid7h) (no related file):*
> In addition, propagate the exit code from a forked process. That use case is actually broken and will be fixed in next commit.

I believe this para in the commit message should be removed (in the final rebase).

---

**Review by mkow — CHANGES_REQUESTED** [review]
> Reviewed 8 of 8 files at r2.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/1943)* status: all files reviewed, 2 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @pwmarcz and @yamahata)

___
*[LibOS/shim/test/regression/exit_group.c, line 53 at r2](https://reviewable.io/reviews/oscarlab/graphene/1943#-MMaBHFADpaBtL-Yz9aC:-MMaBHFADpaBtL-Yz9aD:biy5cnj) ([raw file](https://github.com/oscarlab/graphene/blob/6eb656b11770d8ca26bf12a79f977f25c7c36f89/LibOS/shim/test/regression/exit_group.c#L53)):*
> ```C
> 
>     pid_t pid = fork();
>     if (fork < 0) {
> ```

Ouch. Are you testing here a function pointer for being negative? :P
Really weird that this didn't produce a warning...

**Review by pwmarcz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/1943)* status: 2 of 3 files reviewed, 2 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ), "fixup! " found in commit messages' one-liners (waiting on @dimakuv, @mkow, and @yamahata)

___
*[LibOS/shim/test/regression/exit_group.c, line 53 at r2](https://reviewable.io/reviews/oscarlab/graphene/1943#-MMaBHFADpaBtL-Yz9aC:-MMaGiSD80CpBB1Ne1LV:b-3gk5k0) ([raw file](https://github.com/oscarlab/graphene/blob/6eb656b11770d8ca26bf12a79f977f25c7c36f89/LibOS/shim/test/regression/exit_group.c#L53)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

Ouch. Are you testing here a function pointer for being negative? :P
Really weird that this didn't produce a warning...
</blockquote></details>

Wow :) Yeah, crazy that this compiled without warnings. Fixed.

**Review by mkow — APPROVED** [review]
> Reviewed 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/1943)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: ITL), "fixup! " found in commit messages' one-liners (waiting on @yamahata)

**Review by mkow — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/1943)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: ITL) (waiting on @yamahata)

**Review by dimakuv — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/1943)* status: all files reviewed, 1 unresolved discussion (waiting on @yamahata)

**Review by yamahata — APPROVED** [review]
> Reviewed 2 of 8 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/1943)* status: all files reviewed, 1 unresolved discussion (waiting on @yamahata)

**Review by boryspoplawski — APPROVED** [review]
> Reviewed 2 of 8 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/1943)* status: :shipit: complete! all files reviewed, all discussions resolved

**Pull request merged**
